source/agora/network/NetworkClient.d

```diff
@@ -113,24 +115,51 @@ class NetworkClient
 
     public PublicKey getPublicKey ()
     {
-        return this.attemptRequest!(API.getPublicKey, Throw.Yes)();
+        return this.attemptRequest!(API.getPublicKey, Throw.Yes)(this.api);
     }
 
     /***************************************************************************
 
-        Register the given address as this node's listener
+        Register the node's address to listen for gossiping messages.
 
-        Params:
-            address = the address to register
+        This method is to call the API endpoint, or 'register_listener` which
+        is not declared as one of the `FullNode` API in `FullNode.d` because
+        we must use the `HTTPServerRequest` object to retrieve the address of
+        a client. The `register_address` REST path is registered in the `runNode`
+        function of `Runner.d`.
 
         Throws:
             `Exception` if the request failed.
 
     ***************************************************************************/
 
-    public void registerListener (Address address)
+    public void registerListener ()
     {
-        return this.attemptRequest!(API.registerListener, Throw.Yes)(address);
+        foreach (idx; 0 .. this.max_retries)
+        {
+            try
+            {
+                requestHTTP(format("%s/register_listener", this.address),
+                    (scope req) {
+                        req.method = HTTPMethod.POST;
+                    },
+                    (scope res) {
+                        if (res.statusCode == 200)
+                            return;
+                        else
+                            log.info("Response of {}/register_listener : {}",
+                                this.address, res);
+                    }
+               );
```

Comment:
This is still not exiting early when there was a success. The `return` is inside the delegate, not in the `try/catch`.